---
title: "Mus chr17 inversions"
author: "[gwct](https://gwct.github.io/)"
date: "`r format(Sys.time(), '%m/%d/%Y %H:%M:%S %Z')`"
output:
  #html_document:
  rmdformats::robobook:
    highlight: kate
    includes:
      in_header: '../html-chunks/rmd_header.html'
    df_print: paged
    code_folding: hide
    number_sections: true
    toc_depth: 3
---

```{r setup, warning=FALSE, message=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)

#library(ggplot2)
library(tidyverse)
library(cowplot)
library(egg)
library(kableExtra)
library(zoo)
library(ggbeeswarm)
library(ggsignif)
library(here)
library("ggtree")
library(eulerr)
source(here("docs", "scripts", "lib", "design.r"))

total_spec = 6

window_size = 10
# Window size in kb

max_tree_rank = 3

```

# Introduction

A closer look at some of the inversions:

```{r read-windows, out.width="50%", fig.align = "center", warning=FALSE}

aln_stats_file = here("data", paste0("mm10-", window_size, "kb-window-stats-no-pahari.tsv"))
aln_stats = read.csv(aln_stats_file, header=T, comment.char="#", sep="\t")
aln_stats$window <- gsub('.noanc', '', aln_stats$window)
# Window alignment stats

##########

infile = here("data", paste0("mm10-", window_size, "kb-topo-counts-no-pahari.csv"))
all_windows = read.csv(infile, header=T, comment.char="#")
all_windows = all_windows[order(-all_windows$end), ]
# Window trees file

##########

inversions_file = here("data", "inversions.bed")
inversions = read_tsv(inversions_file, col_names=c("chr", "start", "end", "id"))

inversions %>%
  kable() %>% 
  kable_styling(bootstrap_options=c("striped", "condended", "responsive"), full_width=F, position="center", fixed_thead=T)
# Inversions

##########

dists_file = here("data", "coord-query-no-pahari", "inv-bps-5-Mb.bed.chr17.dists.all")
dists = read_csv(dists_file)
dists$label = NA
if(!is.na(dists$filt)){
  dists$label = ifelse(dists$rf.st == 0, "Same", "Diff")
}
# Tree dists

##########

intersects_file = here("data", "mm10-10kb-windows-inversions.intersect")
intersects = read_tsv(intersects_file, col_names=c("chr", "start", "end", "window", "inv.chr", "inv.start", "inv.end", "inv.overlap", "inv.overlap.bp"))
intersects = intersects %>% filter(inv.overlap.bp > 1) %>% select(window, inv.overlap, inv.overlap.bp)
# Read the windows with inversion intersects labeled

##########

all_windows = left_join(all_windows, intersects, by="window")
all_windows_f = subset(all_windows, filter=="PASS")
dists = left_join(dists, intersects, by="window")
# Add the inversion intersects to the windows and dists dfs

##########

mm_t_sites_file = here("data", "rnd", "sister", "mm10-mus-t.csv")
mm_t_sites = read_csv(mm_t_sites_file, col_names=c("window", "aln.len", "pair.diffs", "x.out", "y.out")) %>%
  mutate(window = gsub(".filter.fa", "", window)) %>%
  mutate(a.pair = pair.diffs / aln.len) %>%
  mutate(d.pair.jc = -(3/4) * (log(1 - ((4/3) * a.pair)))) %>%
  mutate(a.out.x = x.out / aln.len) %>%
  mutate(d.out.x.jc = -(3/4) * (log(1 - ((4/3) * a.out.x)))) %>%
  mutate(a.out.y = y.out / aln.len) %>%
  mutate(d.out.y.jc = -(3/4) * (log(1 - ((4/3) * a.out.y)))) %>%
  mutate(a.out = (a.out.x + a.out.y) / 2) %>%
  mutate(d.out.jc = (d.out.x.jc + d.out.y.jc) / 2) %>%
  mutate(rnd = a.pair / a.out) %>%
  mutate(rnd.log = log10(rnd)) %>%
  mutate(rnd.jc = d.pair.jc / d.out.jc) %>%
  mutate(rnd.jc.log = log10(rnd.jc))
mm_t_quant = quantile(mm_t_sites$rnd.jc, probs=c(0.05), na.rm=T)

spi_spr_sites_file = here("data", "rnd", "sister", "spicilegus-spretus.csv")
spi_spr_sites = read_csv(spi_spr_sites_file, col_names=c("window", "aln.len", "pair.diffs", "x.out", "y.out")) %>%
  mutate(window = gsub(".filter.fa", "", window)) %>%
  mutate(a.pair = pair.diffs / aln.len) %>%
  mutate(d.pair.jc = -(3/4) * (log(1 - ((4/3) * a.pair)))) %>%
  mutate(a.out.x = x.out / aln.len) %>%
  mutate(d.out.x.jc = -(3/4) * (log(1 - ((4/3) * a.out.x)))) %>%
  mutate(a.out.y = y.out / aln.len) %>%
  mutate(d.out.y.jc = -(3/4) * (log(1 - ((4/3) * a.out.y)))) %>%
  mutate(a.out = (a.out.x + a.out.y) / 2) %>%
  mutate(d.out.jc = (d.out.x.jc + d.out.y.jc) / 2) %>%
  mutate(rnd = a.pair / a.out) %>%
  mutate(rnd.log = log10(rnd)) %>%
  mutate(rnd.jc = d.pair.jc / d.out.jc) %>%
  mutate(rnd.jc.log = log10(rnd.jc))
spi_spr_quant = quantile(spi_spr_sites$rnd.jc, probs=c(0.05), na.rm=T)

mm_spi_sites_file = here("data", "rnd", "non-sister", "mm10-spicilegus.csv")
mm_spi_sites = read_csv(mm_spi_sites_file, col_names=c("window", "aln.len", "pair.diffs", "x.out", "y.out")) %>%
  mutate(window = gsub(".filter.fa", "", window)) %>%
  mutate(a.pair = pair.diffs / aln.len) %>%
  mutate(d.pair.jc = -(3/4) * (log(1 - ((4/3) * a.pair)))) %>%
  mutate(a.out.x = x.out / aln.len) %>%
  mutate(d.out.x.jc = -(3/4) * (log(1 - ((4/3) * a.out.x)))) %>%
  mutate(a.out.y = y.out / aln.len) %>%
  mutate(d.out.y.jc = -(3/4) * (log(1 - ((4/3) * a.out.y)))) %>%
  mutate(a.out = (a.out.x + a.out.y) / 2) %>%
  mutate(d.out.jc = (d.out.x.jc + d.out.y.jc) / 2) %>%
  mutate(rnd = a.pair / a.out) %>%
  mutate(rnd.log = log10(rnd)) %>%
  mutate(rnd.jc = d.pair.jc / d.out.jc) %>%
  mutate(rnd.jc.log = log10(rnd.jc))
mm_spi_quant = quantile(mm_spi_sites$rnd.jc, probs=c(0.05), na.rm=T)

t_spi_sites_file = here("data", "rnd", "non-sister", "mm10-spicilegus.csv")
t_spi_sites = read_csv(t_spi_sites_file, col_names=c("window", "aln.len", "pair.diffs", "x.out", "y.out")) %>%
  mutate(window = gsub(".filter.fa", "", window)) %>%
  mutate(a.pair = pair.diffs / aln.len) %>%
  mutate(d.pair.jc = -(3/4) * (log(1 - ((4/3) * a.pair)))) %>%
  mutate(a.out.x = x.out / aln.len) %>%
  mutate(d.out.x.jc = -(3/4) * (log(1 - ((4/3) * a.out.x)))) %>%
  mutate(a.out.y = y.out / aln.len) %>%
  mutate(d.out.y.jc = -(3/4) * (log(1 - ((4/3) * a.out.y)))) %>%
  mutate(a.out = (a.out.x + a.out.y) / 2) %>%
  mutate(d.out.jc = (d.out.x.jc + d.out.y.jc) / 2) %>%
  mutate(rnd = a.pair / a.out) %>%
  mutate(rnd.log = log10(rnd)) %>%
  mutate(rnd.jc = d.pair.jc / d.out.jc) %>%
  mutate(rnd.jc.log = log10(rnd.jc))
t_spi_quant = quantile(t_spi_sites$rnd.jc, probs=c(0.05), na.rm=T)

```


```{r topo-counts, out.width="50%", fig.align = "center", warning=FALSE}

topo_counts = unique(subset(all_windows_f, select=c(topo.num.overall, topo.count.overall, topo.rank.overall, topo.color)))
topo_counts = topo_counts[order(topo_counts$topo.rank.overall),]
num_topos = nrow(topo_counts)
# Get the topology counts

#print(paste0("In total ", num_topos, " different topologies were recovered"))
# Display some info

col_rank = 1
for(i in 1:nrow(topo_counts)){
  row = topo_counts[i,]
  cur_col = row$topo.color
  if(cur_col != "#999999"){
    new_label = paste("Top ", col_rank, sep="")
    col_rank = col_rank + 1
  }else{
    new_label = "Other"
  }
  topo_counts$col.cat[topo_counts$topo.num.overall==i] = new_label
  
  if(row$topo.rank.overall > max_tree_rank){
    topo_counts[i,]$topo.color = "#333333"
  }
}
# Add label and color columns to the topo counts

topo_counts$outline = "transparent"
# Add an outline column -- black outline is chrome topo

topo_counts$topo.prop = round(topo_counts$topo.count / sum(topo_counts$topo.count), 2)
# Percentage for each topology.

```


```{r gen-chromoplot, out.width="65%", fig.align = "center", warning=FALSE, fig.height=4}

genChromoPlot <- function(chrome_str, label, features=FALSE, cur_feature=FALSE, num_flank_windows=0, multi_row=TRUE){
# chrome_str:       the chromosome to be subset
# label:            a title, not currently used
# features:         a set of coordinates to be marked on the chromoplot
# cur_feature:      if a single feature is provided here, the chromoplot will be subset to show just that region
# num_flank_window: the number of windows surrounding a feature to subset; to be used with cur_feature, otherwise ignored
# multi_row:        whether or not to display the top three topologies in separate rows on the chromoplot
  
  if(!cur_feature==FALSE){
    flank = num_flank_windows * window_size * 1000
    chrdata = subset(all_windows, chr==chrome_str & start > cur_feature$start - flank & end < cur_feature$end + flank)
  }else{
    chrdata = subset(all_windows, chr==chrome_str)
  }
  # Subset the data by chromosome and feature (if provided)
  
  ##########
  
  chr_len = chrdata[chrdata$chr==chrome_str, ]$chr.len[1]
  # Get the chromosome length
  
  chrdata$col.cat = NA
  chrdata$ystart = NA
  chrdata$yend = NA
  # Initialize some empty columns in the data
  
  cur_col = "#ffffff"
  topo_counts$topo.color = as.character(topo_counts$topo.color)
  # Other setup
  
  for(i in 1:nrow(chrdata)) {
    cur_topo_num = chrdata[i,]$topo.num.overall
    # Get the topology number/id of the current window
    
    if(aln_stats[aln_stats$window == chrdata[i,]$window,]$total.seqs.filter != 5){
      cur_col = "#999999"
      chrdata[i,]$ystart = 0
      chrdata[i,]$yend = 4
    # If this window was filtered out in the alignment steps, set the color to grey and the y coords to span the chromoplot (0 to 4)
    }else if(is.na(cur_topo_num)){
      cur_col = "#999999"
      chrdata[i,]$ystart = 0
      chrdata[i,]$yend = 4
    # If this window was filtered out otherwise, set the color to grey and the y coords to span the chromoplot (0 to 4)
    }else{
      cur_col = topo_counts$topo.color[topo_counts$topo.num.overall==cur_topo_num]
    # If this window was not filtered, set the current color to be the one for the rank by frequency
    }
    
    chrdata[i,]$col.cat = cur_col
    # Set the color for this window based on above
  }
  ## Categorize the windows and get their colors
  
  ##########
  
  if(multi_row){
    chrdata$ystart[chrdata$topo.rank.overall==1] = 4
    chrdata$yend[chrdata$topo.rank.overall==1] = 3
    
    chrdata$ystart[chrdata$topo.rank.overall==2] = 3
    chrdata$yend[chrdata$topo.rank.overall==2] = 2
    
    chrdata$ystart[chrdata$topo.rank.overall==3] = 2
    chrdata$yend[chrdata$topo.rank.overall==3] = 1
    
    chrdata$ystart[chrdata$topo.rank.overall>3] = 1
    chrdata$yend[chrdata$topo.rank.overall>3] = 0
  # If the plot is multi-line, topologies of rank 1, 2, and 3 get their own rows
  }else{
    chrdata$ystart = 1
    chrdata$yend = 0
  # If the plot is not multi-line, all columns go from 0 to 1
  }
  ## Set the y coordinates for the chromoplot
  
  ##########
  
  cols_labs = levels(as.factor(chrdata$col.cat))
  names(cols_labs) = levels(as.factor(chrdata$col.cat))
  # Set up the colors
  
  cur_len = max(chrdata$end) - min(chrdata$start)
  chr_breaks = seq(min(chrdata$start), max(chrdata$end), by=cur_len/5)
  # Set up the x-axis breaks
  
  chromoplot = ggplot(chrdata, aes(x=start, y=ystart, color=col.cat, fill=col.cat))
  
  if(!features==FALSE){
    offset = 3
    feature_cols = corecol(numcol=nrow(features), offset=offset)
    chromoplot = chromoplot +
      geom_rect(data=features, inherit.aes=F, aes(ymin=-Inf, ymax=Inf, xmin=start, xmax=end), alpha=0.2, fill=feature_cols) 
  }
    
  chromoplot = chromoplot + 
    geom_rect(aes(ymin=ystart, ymax=yend, xmin=start, xmax=end)) +
    scale_color_manual(values=cols_labs) +
    scale_fill_manual(values=cols_labs) +
    scale_x_continuous(limits=c(min(chrdata$start), max(chrdata$end)), breaks=chr_breaks) +
    xlab("Position") +
    bartheme() +
    theme(legend.position="none",
          #legend.key.width = unit(0.75,  unit = "cm"),
          #legend.spacing.x = unit(0.25, 'cm'),
          #legend.title = element_blank(),
          #legend.text=element_text(size=12),
          axis.text=element_text(size=8),
          axis.title=element_text(size=10),
          axis.line.y=element_line(color="white"),
          axis.ticks.y=element_line(color="white"),
          axis.text.y=element_text(color="white"),
          axis.title.y=element_text(angle=0),
          plot.margin = unit(c(0,0,0,0), "cm")
  )

  if(!features==FALSE){
    offset = 3
    for(i in 1:nrow(features)){
      chromoplot = chromoplot +
        geom_vline(xintercept=features[i,]$start, color=corecol(numcol=1, offset=offset), size=1) +
        geom_vline(xintercept=features[i,]$end, color=corecol(numcol=1, offset=offset), size=1)
      offset = offset + 1
    }
  }
  # If a set of features have been provided to be marked, draw vertical lines at their boundaries here
  
  if(multi_row){
    chromoplot = chromoplot + 
      scale_y_continuous(limits=c(0,4), breaks=0.5:3.5, labels=c("0.0", "3", "2", "1")) +
      ylab("Topology")
  }else{
    chromoplot = chromoplot + 
      scale_y_continuous(limits=c(0,1), breaks=0.5:1.5, labels=c("0.0", "")) +
      ylab("Topology")
  }
  # Set up the y-axis based on whether the plot is multi-line or not
  
  return(chromoplot)
  # Return the plot
}


```


```{r gen-rfplot, out.width="85%", fig.align = "center", warning=FALSE}

genRFPlot <- function(df, yvar, chrome_str, label, ylab, ymax=FALSE, quantline=FALSE, features=FALSE, cur_feature=FALSE, num_flank_windows=0){
# chrome_str:       the chromosome to be subset
# label:            a title, not currently used
# features:         a set of coordinates to be marked on the chromoplot
# cur_feature:      if a single feature is provided here, the chromoplot will be subset to show just that region
# num_flank_window: the number of windows surrounding a feature to subset; to be used with cur_feature, otherwise ignored
  
  if(!cur_feature==FALSE){
    flank = num_flank_windows * window_size * 1000
    chrdata = subset(df, chr==chrome_str & start > cur_feature$start - flank & end < cur_feature$end + flank)
  }else{
    chrdata = subset(df, chr==chrome_str)
  }
  # Subset the data by chromosome and feature (if provided)
  
  ##########
  
  #qts = quantile(chrdata$wrf.st, probs=c(.05,.95), na.rm=T)
  # Calculate some percentiles (not currently used)
  
  cols = rev(c(corecol(pal="wilke", numcol=4, offset=4)))
  # Set up the colors (not currently used)

  cur_len = max(chrdata$end) - min(chrdata$start)
  chr_breaks = seq(min(chrdata$start), max(chrdata$end), by=cur_len/5)
  # Set up the x-axis breaks
  
  wrf_adj = ggplot(chrdata, aes_string(x="start", y=yvar))
  
  if(!features==FALSE){
    offset = 3
    feature_cols = corecol(numcol=nrow(features), offset=offset)
    wrf_adj = wrf_adj +
      geom_rect(data=features, inherit.aes=F, aes(ymin=0, ymax=Inf, xmin=start, xmax=end), alpha=0.2, fill=feature_cols)
  }
  
  wrf_adj = wrf_adj +
    geom_point(size=1, alpha=0.2) +
    #geom_line(aes(y=rollmean(chrdata[[yvar]], 5, na.pad=T))) +
    #geom_line() +
    scale_x_continuous(limits=c(min(chrdata$start), max(chrdata$end)), breaks=chr_breaks)
  
  if(!quantline==FALSE){
    wrf_adj = wrf_adj +
      geom_hline(yintercept=quantline, size=0.75, linetype="dashed", color="#f0e442")
  }
  
  if(!ymax==FALSE){
    if(ymax=="logy"){
      wrf_adj = wrf_adj +
        scale_y_log10()
    }else{
      wrf_adj = wrf_adj +
        scale_y_continuous(limits=c(0, ymax))
    }
  }

  wrf_adj = wrf_adj +
    xlab(NULL) +
    ylab(ylab) +
    #scale_color_manual(values=cols) +
    #scale_fill_manual(values=c("Same topology"="#999999", "Diff. topology"="#920000")) +
    bartheme() +
    #theme_classic() +
    theme(axis.text=element_text(size=8),
          axis.title=element_text(size=10),
          #axis.line.x=element_blank(),
          axis.ticks.x=element_blank(),
          axis.title.x=element_blank(),
          axis.text.x=element_blank(),
          axis.title.y=element_text(angle=0),
          plot.margin = grid::unit(c(0,0,0,0), "cm"),
          axis.ticks.length.x = unit(2, "pt"))
          #plot.background = element_rect(color = "deepskyblue3", linewidth = 3))
    
  if(!features==FALSE){
    offset = 3
    for(i in 1:nrow(features)){
      wrf_adj = wrf_adj +
        geom_vline(xintercept=features[i,]$start, color=corecol(numcol=1, offset=offset), size=1) +
        geom_vline(xintercept=features[i,]$end, color=corecol(numcol=1, offset=offset), size=1)
      offset = offset + 1
    }
  }
  # If a set of features have been provided to be marked, draw vertical lines at their boundaries here
  
  return(wrf_adj)
  # Return the plot
}

```

In the plots below, I show several statistics on 10kb windows along mm10 chromosome 17 and outline the boundaries of the inversions. I show:

1. For two branches, the divergence of each pair of sequences, $d$ (or $d_{XY}$), under a simple Jukes-Cantor model of nucleotide substitution. Given that for any pair of sequences: $$ p = \frac{n_d}{n} $$ where $n_d$ is the number of sites where the nucleotides differ between the two sequences and $n$ is the total number of aligned (non-gap) sites between the two sequences. Then: $$ d_{XY} = -\frac{3}{4} \ln(1 - \frac{4}{3}p) $$

2. Relative Node Depth ($RND$) for the specified pair of sequences. This statistic uses $d$ between a pair of species, $d_{XY}$, and compares it to the average $d$ from each species to an outgroup, $d_{out}$. $d_{XY}$ is calculated as above, while $d_{XO}$ and $d_{YO}$ are calculated the same way, but between species $X$ and the outgroup, $O$, and species $Y$ and $O$. Then, $$ d_{out} = \frac{d_{XO} +d_{YO}}{2} $$ and $$ RND = \frac{d_{XY}}{d_{out}} $$ LOWER values of $RND$ can be indicative of introgression. See [Rosenzweig et al. 2016](https://hahnlab.sitehost.iu.edu/Publications/Rosenzweig_etal2016.pdf) for more info.

3. Weighted Robinson-Foulds (wRF) distance between each window tree and the SPECIES TREE. I gave a brief explanation of wRF [here](https://gwct.github.io/mus-t-haplotype/trees_no_pahari.html#tree-similarity-in-windows-around-inversion-breakpoints). In this case, the REFERENCE TREE is ALWAYS the SPECIES TREE and the QUERY TREE is the tree inferred from that 10kb window.

4. The chromoplot visualizing topological differences in window trees. The colors are the same as [here](https://gwct.github.io/mus-t-haplotype/trees_no_pahari.html#distribution-of-tree-topologies).

For each panel (except the chromoplot of topologies), the dots represent the statistic for a single 10kb window. For the $RND$ plots, the dashed, yellow line represents the 5th percentile for ALL windows across the chromosome. Even in the individual inversion plots, this yellow line is for the whole chromosome.

The $RND$ values are log10 scaled.

For each panel, vertical shaded regions represent the locations of the inversions.

# Chromosome 17

```{r combine, out.width="75%", fig.align = "center", warning=FALSE, fig.height=4}

cur_chromoplot = genChromoPlot("chr17", "Chromosome 17", features=inversions)

#####

cur_rfplot = genRFPlot(dists, "wrf.st", "chr17", "Chromosome 17", "wRF to\nspecies tree", 0.5, features=inversions)

#####

mm_t_d_plot = genRFPlot(left_join(all_windows, mm_t_sites, by="window"), "d.pair.jc", "chr17", "Chromosome 17", expression(paste("mm10-mus-t ", italic("d"))), 1.0, features=inversions)

mm_t_rnd_plot = genRFPlot(left_join(all_windows, mm_t_sites, by="window"), "rnd.jc", "chr17", "Chromosome 17", "mm10-mus-t\nlog(RND)",  "logy", mm_t_quant, features=inversions)

#####

spi_spr_d_plot = genRFPlot(left_join(all_windows, spi_spr_sites, by="window"), "d.pair.jc", "chr17", "Chromosome 17", expression(atop("spicilegus-spretus")), 1.0, features=inversions)

spi_spr_rnd_plot = genRFPlot(left_join(all_windows, spi_spr_sites, by="window"), "rnd.jc", "chr17", "Chromosome 17", "spicilegus-spretus\nlog(RND)", "logy", spi_spr_quant, features=inversions)

#####

mm_spi_d_plot = genRFPlot(left_join(all_windows, mm_spi_sites, by="window"), "d.pair.jc", "chr17", "Chromosome 17", expression(paste("mm10-spicilegus ", italic("d"))), 1.0, features=inversions)

mm_spi_rnd_plot = genRFPlot(left_join(all_windows, mm_spi_sites, by="window"), "rnd.jc", "chr17", "Chromosome 17", "mm10-spicilegus\nlog(RND)",  "logy", mm_spi_quant, features=inversions)

#####

t_spi_d_plot = genRFPlot(left_join(all_windows, t_spi_sites, by="window"), "d.pair.jc", "chr17", "Chromosome 17", expression(paste("mus-t-spicilegus ", italic("d"))), 1.0, features=inversions)

t_spi_rnd_plot = genRFPlot(left_join(all_windows, t_spi_sites, by="window"), "rnd.jc", "chr17", "Chromosome 17", "mus-t-spicilegus\nlog(RND)",  "logy", t_spi_quant, features=inversions)

#####

p = plot_grid(mm_spi_d_plot, mm_spi_rnd_plot, t_spi_d_plot, t_spi_rnd_plot, cur_rfplot, cur_chromoplot, nrow=6, align="v", rel_heights=c(1,1,1,1,1,2))
print(p)

```

# Inversion 1

```{r inv1, out.width="75%", fig.align = "center", warning=FALSE, fig.height=4}

cur_chromoplot = genChromoPlot("chr17", "Inversion 1", features=inversions, cur_feature=inversions[1,], num_flank_windows=50, multi_row=F)

#####

cur_rfplot = genRFPlot(dists, "wrf.st", "chr17", "Inversion 1", "wRF to\nspecies tree", 0.5, features=inversions, cur_feature=inversions[1,], num_flank_windows=50)


#####

mm_t_d_plot = genRFPlot(left_join(all_windows, mm_t_sites, by="window"), "d.pair.jc", "chr17", "Inversion 1", expression(paste("mm10-mus-t ", italic("d"))), 1.0, features=inversions, cur_feature=inversions[1,], num_flank_windows=50)

mm_t_rnd_plot = genRFPlot(left_join(all_windows, mm_t_sites, by="window"), "rnd.jc", "chr17", "Inversion 1", "mm10-mus-t\nlog(RND)",  "logy", mm_t_quant, features=inversions, cur_feature=inversions[1,], num_flank_windows=50)

#####

spi_spr_d_plot = genRFPlot(left_join(all_windows, spi_spr_sites, by="window"), "d.pair.jc", "chr17", "Inversion 1", expression(atop("spicilegus-spretus")), 1.0, features=inversions, cur_feature=inversions[1,], num_flank_windows=50)

spi_spr_rnd_plot = genRFPlot(left_join(all_windows, spi_spr_sites, by="window"), "rnd.jc", "chr17", "Inversion 1", "spicilegus-spretus\nlog(RND)", "logy", spi_spr_quant, features=inversions, cur_feature=inversions[1,], num_flank_windows=50)

#####

mm_spi_d_plot = genRFPlot(left_join(all_windows, mm_spi_sites, by="window"), "d.pair.jc", "chr17", "Inversion 1", expression(paste("mm10-spicilegus ", italic("d"))), 1.0, features=inversions, cur_feature=inversions[1,], num_flank_windows=50)

mm_spi_rnd_plot = genRFPlot(left_join(all_windows, mm_spi_sites, by="window"), "rnd.jc", "chr17", "Inversion 1", "mm10-spicilegus\nlog(RND)",  "logy", mm_spi_quant, features=inversions, cur_feature=inversions[1,], num_flank_windows=50)

#####

t_spi_d_plot = genRFPlot(left_join(all_windows, t_spi_sites, by="window"), "d.pair.jc", "chr17", "Inversion 1", expression(paste("mus-t-spicilegus ", italic("d"))), 1.0, features=inversions, cur_feature=inversions[1,], num_flank_windows=50)

t_spi_rnd_plot = genRFPlot(left_join(all_windows, t_spi_sites, by="window"), "rnd.jc", "chr17", "Inversion 1", "mus-t-spicilegus\nlog(RND)",  "logy", t_spi_quant, features=inversions, cur_feature=inversions[1,], num_flank_windows=50)

#####

p = plot_grid(mm_spi_d_plot, mm_spi_rnd_plot, t_spi_d_plot, t_spi_rnd_plot, cur_rfplot, cur_chromoplot, nrow=6, align="v", rel_heights=c(1,1,1,1,1,2))
print(p)

```

# Inversion 2

```{r inv2, out.width="75%", fig.align = "center", warning=FALSE, fig.height=4}

cur_chromoplot = genChromoPlot("chr17", "Inversion 2", features=inversions, cur_feature=inversions[2,], num_flank_windows=50, multi_row=F)

#####

cur_rfplot = genRFPlot(dists, "wrf.st", "chr17", "Inversion 2", "wRF to\nspecies tree", 0.5, features=inversions, cur_feature=inversions[2,], num_flank_windows=50)


#####

mm_t_d_plot = genRFPlot(left_join(all_windows, mm_t_sites, by="window"), "d.pair.jc", "chr17", "Inversion 2", expression(paste("mm10-mus-t ", italic("d"))), 1.0, features=inversions, cur_feature=inversions[2,], num_flank_windows=50)

mm_t_rnd_plot = genRFPlot(left_join(all_windows, mm_t_sites, by="window"), "rnd.jc", "chr17", "Inversion 2", "mm10-mus-t\nlog(RND)",  "logy", mm_t_quant, features=inversions, cur_feature=inversions[2,], num_flank_windows=50)

#####

spi_spr_d_plot = genRFPlot(left_join(all_windows, spi_spr_sites, by="window"), "d.pair.jc", "chr17", "Inversion 2", expression(atop("spicilegus-spretus")), 1.0, features=inversions, cur_feature=inversions[2,], num_flank_windows=50)

spi_spr_rnd_plot = genRFPlot(left_join(all_windows, spi_spr_sites, by="window"), "rnd.jc", "chr17", "Inversion 2", "spicilegus-spretus\nlog(RND)", "logy", spi_spr_quant, features=inversions, cur_feature=inversions[2,], num_flank_windows=50)

#####

mm_spi_d_plot = genRFPlot(left_join(all_windows, mm_spi_sites, by="window"), "d.pair.jc", "chr17", "Inversion 2", expression(paste("mm10-spicilegus ", italic("d"))), 1.0, features=inversions, cur_feature=inversions[2,], num_flank_windows=50)

mm_spi_rnd_plot = genRFPlot(left_join(all_windows, mm_spi_sites, by="window"), "rnd.jc", "chr17", "Inversion 2", "mm10-spicilegus\nlog(RND)",  "logy", mm_spi_quant, features=inversions, cur_feature=inversions[2,], num_flank_windows=50)

#####

t_spi_d_plot = genRFPlot(left_join(all_windows, t_spi_sites, by="window"), "d.pair.jc", "chr17", "Inversion 2", expression(paste("mus-t-spicilegus ", italic("d"))), 1.0, features=inversions, cur_feature=inversions[2,], num_flank_windows=50)

t_spi_rnd_plot = genRFPlot(left_join(all_windows, t_spi_sites, by="window"), "rnd.jc", "chr17", "Inversion 2", "mus-t-spicilegus\nlog(RND)",  "logy", t_spi_quant, features=inversions, cur_feature=inversions[2,], num_flank_windows=50)

#####

p = plot_grid(mm_spi_d_plot, mm_spi_rnd_plot, t_spi_d_plot, t_spi_rnd_plot, cur_rfplot, cur_chromoplot, nrow=6, align="v", rel_heights=c(1,1,1,1,1,2))
print(p)

```

# Inversion 3

```{r inv3, out.width="75%", fig.align = "center", warning=FALSE, fig.height=4}

cur_chromoplot = genChromoPlot("chr17", "Inversion 3", features=inversions, cur_feature=inversions[3,], num_flank_windows=50, multi_row=F)

#####

cur_rfplot = genRFPlot(dists, "wrf.st", "chr17", "Inversion 3", "wRF to\nspecies tree", 0.5, features=inversions, cur_feature=inversions[3,], num_flank_windows=50)


#####

mm_t_d_plot = genRFPlot(left_join(all_windows, mm_t_sites, by="window"), "d.pair.jc", "chr17", "Inversion 3", expression(paste("mm10-mus-t ", italic("d"))), 1.0, features=inversions, cur_feature=inversions[3,], num_flank_windows=50)

mm_t_rnd_plot = genRFPlot(left_join(all_windows, mm_t_sites, by="window"), "rnd.jc", "chr17", "Inversion 3", "mm10-mus-t\nlog(RND)",  "logy", mm_t_quant, features=inversions, cur_feature=inversions[3,], num_flank_windows=50)

#####

spi_spr_d_plot = genRFPlot(left_join(all_windows, spi_spr_sites, by="window"), "d.pair.jc", "chr17", "Inversion 3", expression(atop("spicilegus-spretus")), 1.0, features=inversions, cur_feature=inversions[3,], num_flank_windows=50)

spi_spr_rnd_plot = genRFPlot(left_join(all_windows, spi_spr_sites, by="window"), "rnd.jc", "chr17", "Inversion 3", "spicilegus-spretus\nlog(RND)", "logy", spi_spr_quant, features=inversions, cur_feature=inversions[3,], num_flank_windows=50)

#####

mm_spi_d_plot = genRFPlot(left_join(all_windows, mm_spi_sites, by="window"), "d.pair.jc", "chr17", "Inversion 3", expression(paste("mm10-spicilegus ", italic("d"))), 1.0, features=inversions, cur_feature=inversions[3,], num_flank_windows=50)

mm_spi_rnd_plot = genRFPlot(left_join(all_windows, mm_spi_sites, by="window"), "rnd.jc", "chr17", "Inversion 3", "mm10-spicilegus\nlog(RND)",  "logy", mm_spi_quant, features=inversions, cur_feature=inversions[3,], num_flank_windows=50)

#####

t_spi_d_plot = genRFPlot(left_join(all_windows, t_spi_sites, by="window"), "d.pair.jc", "chr17", "Inversion 3", expression(paste("mus-t-spicilegus ", italic("d"))), 1.0, features=inversions, cur_feature=inversions[3,], num_flank_windows=50)

t_spi_rnd_plot = genRFPlot(left_join(all_windows, t_spi_sites, by="window"), "rnd.jc", "chr17", "Inversion 3", "mus-t-spicilegus\nlog(RND)",  "logy", t_spi_quant, features=inversions, cur_feature=inversions[3,], num_flank_windows=50)

#####

p = plot_grid(mm_spi_d_plot, mm_spi_rnd_plot, t_spi_d_plot, t_spi_rnd_plot, cur_rfplot, cur_chromoplot, nrow=6, align="v", rel_heights=c(1,1,1,1,1,2))
print(p)

```

# Inversion 4

```{r inv4, out.width="75%", fig.align = "center", warning=FALSE, fig.height=4}

cur_chromoplot = genChromoPlot("chr17", "Inversion 4", features=inversions, cur_feature=inversions[4,], num_flank_windows=50, multi_row=F)

#####

cur_rfplot = genRFPlot(dists, "wrf.st", "chr17", "Inversion 4", "wRF to\nspecies tree", 0.5, features=inversions, cur_feature=inversions[4,], num_flank_windows=50)


#####

mm_t_d_plot = genRFPlot(left_join(all_windows, mm_t_sites, by="window"), "d.pair.jc", "chr17", "Inversion 4", expression(paste("mm10-mus-t ", italic("d"))), 1.0, features=inversions, cur_feature=inversions[4,], num_flank_windows=50)

mm_t_rnd_plot = genRFPlot(left_join(all_windows, mm_t_sites, by="window"), "rnd.jc", "chr17", "Inversion 4", "mm10-mus-t\nlog(RND)",  "logy", mm_t_quant, features=inversions, cur_feature=inversions[4,], num_flank_windows=50)

#####

spi_spr_d_plot = genRFPlot(left_join(all_windows, spi_spr_sites, by="window"), "d.pair.jc", "chr17", "Inversion 4", expression(atop("spicilegus-spretus")), 1.0, features=inversions, cur_feature=inversions[4,], num_flank_windows=50)

spi_spr_rnd_plot = genRFPlot(left_join(all_windows, spi_spr_sites, by="window"), "rnd.jc", "chr17", "Inversion 4", "spicilegus-spretus\nlog(RND)", "logy", spi_spr_quant, features=inversions, cur_feature=inversions[4,], num_flank_windows=50)

#####

mm_spi_d_plot = genRFPlot(left_join(all_windows, mm_spi_sites, by="window"), "d.pair.jc", "chr17", "Inversion 4", expression(paste("mm10-spicilegus ", italic("d"))), 1.0, features=inversions, cur_feature=inversions[4,], num_flank_windows=50)

mm_spi_rnd_plot = genRFPlot(left_join(all_windows, mm_spi_sites, by="window"), "rnd.jc", "chr17", "Inversion 4", "mm10-spicilegus\nlog(RND)",  "logy", mm_spi_quant, features=inversions, cur_feature=inversions[4,], num_flank_windows=50)

#####

t_spi_d_plot = genRFPlot(left_join(all_windows, t_spi_sites, by="window"), "d.pair.jc", "chr17", "Inversion 4", expression(paste("mus-t-spicilegus ", italic("d"))), 1.0, features=inversions, cur_feature=inversions[4,], num_flank_windows=50)

t_spi_rnd_plot = genRFPlot(left_join(all_windows, t_spi_sites, by="window"), "rnd.jc", "chr17", "Inversion 4", "mus-t-spicilegus\nlog(RND)",  "logy", t_spi_quant, features=inversions, cur_feature=inversions[4,], num_flank_windows=50)

#####

p = plot_grid(mm_spi_d_plot, mm_spi_rnd_plot, t_spi_d_plot, t_spi_rnd_plot, cur_rfplot, cur_chromoplot, nrow=6, align="v", rel_heights=c(1,1,1,1,1,2))
print(p)

```






